File: app/src/main/java/com/example/foodorderingapp/adapter/CartAdapter.java
package com.example.foodorderingapp.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.example.foodorderingapp.databinding.CartItemBinding;

import java.util.List;

public class CartAdapter extends RecyclerView.Adapter<CartAdapter.CartViewHolder> {
    private final List<String> cartItems;
    private final List<String> cartItemsPrice;
    private final List<Integer> cartImages;
    private final int[] itemQuantities;

    public CartAdapter(List<String> cartItems, List<String> cartItemsPrice, List<Integer> cartImages) {
        this.cartItems = cartItems;
        this.cartItemsPrice = cartItemsPrice;
        this.cartImages = cartImages;
        this.itemQuantities = new int[cartImages.size()];
        for (int i = 0; i < itemQuantities.length; i++) {
            itemQuantities[i] = 1; // Initialize quantities to 1
        }
    }

    @NonNull
    @Override
    public CartViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        CartItemBinding binding = CartItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new CartViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull CartViewHolder holder, int position) {
        holder.bind(position);
    }

    @Override
    public int getItemCount() {
        return cartItems.size();
    }

    class CartViewHolder extends RecyclerView.ViewHolder {
        private final CartItemBinding binding;

        public CartViewHolder(CartItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(int position) {
            binding.cartFoodName.setText(cartItems.get(position));
            binding.cartItemPrice.setText(cartItemsPrice.get(position));
            binding.cartImage.setImageResource(cartImages.get(position));
            binding.quantity.setText(String.valueOf(itemQuantities[position]));

            binding.minusButton.setOnClickListener(v -> decreaseQuantity(position));
            binding.plusButton.setOnClickListener(v -> increaseQuantity(position));
            binding.deleteButton.setOnClickListener(v -> {
                int itemPosition = getAdapterPosition();
                if (itemPosition != RecyclerView.NO_POSITION) {
                    deleteItem(itemPosition);
                }
            });
        }

        private void decreaseQuantity(int position) {
            if (itemQuantities[position] > 1) {
                itemQuantities[position]--;
                binding.quantity.setText(String.valueOf(itemQuantities[position]));
            }
        }

        private void increaseQuantity(int position) {
            if (itemQuantities[position] < 10) {
                itemQuantities[position]++;
                binding.quantity.setText(String.valueOf(itemQuantities[position]));
            }
        }

        private void deleteItem(int position) {
            cartItems.remove(position);
            cartItemsPrice.remove(position);
            cartImages.remove(position);
            notifyItemRemoved(position);
            notifyItemRangeChanged(position, cartItems.size());
        }
    }
}

File: app/src/main/java/com/example/foodorderingapp/adapter/CartAdapter.kt
package com.example.foodorderingapp.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Recycler
import com.example.foodorderingapp.databinding.CartItemBinding

class CartAdapter(
    private val cartItems: MutableList<String>,
    private val cartItemsPrice: MutableList<String>,
    private val cartImages: MutableList<Int>
) : RecyclerView.Adapter<CartAdapter.CartViewHolder>() {
    private val itemQuantities = IntArray(cartImages.size){1}

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): CartViewHolder {
        val binding = CartItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CartViewHolder(binding)
    }

    override fun onBindViewHolder(
        holder: CartViewHolder,
        position: Int,
    ) {
        holder.bind(position)
    }

    override fun getItemCount(): Int = cartItems.size

    inner class CartViewHolder(
        private val binding: CartItemBinding,
    ) : RecyclerView.ViewHolder(binding.root) {
        fun bind(position: Int) {
            binding.apply {
                val itemquantity = itemQuantities[position]
                cartFoodName.text = cartItems[position]
                cartItemPrice.text = cartItemsPrice[position]
                cartImage.setImageResource(cartImages[position])
                quantity.text = itemquantity.toString()

                minusButton.setOnClickListener {
        decreaseQuantity(position)
                }

                plusButton.setOnClickListener {
  increaseQuantity(position)
                }
            }
        }

        private fun decreaseQuantity(position: Int) {
            if (itemQuantities[position] >1) {
                itemQuantities[position]--
                binding.quantity.text = itemQuantities[position].toString()
            }

            if(itemQuantities[position] <= 1) {
                val itemPosition = adapterPosition
                if (itemPosition!= RecyclerView.NO_POSITION) {
                    deleteItem(itemPosition)
                }
            }
        }

        fun increaseQuantity(position: Int) {
            if (itemQuantities[position] <10) {
                itemQuantities[position]++
                binding.quantity.text = itemQuantities[position].toString()
            }
        }

        fun deleteItem(position: Int) {
            cartItems.removeAt(position)
            cartItemsPrice.removeAt(position)
            cartImages.removeAt(position)
            notifyItemRemoved(position)
            notifyItemRangeChanged(position, cartItems.size)
        }
    }
}
